.github/workflows/build.yml

```diff
@@ -0,0 +1,141 @@
+
+  Dist:
+    runs-on: ubuntu-latest
+    needs:
+      - Verify
+
+    env:
+      GOPATH: /home/runner/go
+
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        fetch-depth: 0
+
+    - name: Set up Go
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.16
+
+    - name: Cache Godel assets
+      uses: actions/cache@v2
+      with:
+        path: ~/.godel
+        key: ${{ runner.os }}-godel-${{ hashFiles('godelw', 'godel/config/godel.yml') }}
+        restore-keys: |
+          ${{ runner.os }}-godel-
+
+    - name: Set up Node
+      uses: actions/setup-node@v2
+      with:
+        node-version: 12
+
+    - name: Cache node_modules
+      uses: actions/cache@v2
+      with:
+        path: node_modules
+        key: ${{ runner.os }}-node-modules-${{ hashFiles('yarn.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-node-modules-
+
+    - name: Install frontend dependencies
+      run: yarn install
+
+    - name: Build frontend
+      run: yarn run build:production
+
+    - name: Build distribution
+      run: ./godelw dist
+
+    - name: Build Docker image
+      run: ./godelw docker build --verbose
+
+    - name: Archive distribution
+      uses: actions/upload-artifact@v2
+      with:
+        name: dist
+        path: |
+          build/policy-bot/*/bin/*.tgz
+
+    #
+    # Steps after this point should only run when publishing
+    # Include them here to avoid exporting the Docker container as an artifact
+    #
+
+    - name: Login to Docker Hub
+      if: ${{ github.event_name == 'push' }}
+      uses: docker/login-action@v1
+      with:
+        username: ${{ secrets.DOCKERHUB_USERNAME }}
+        password: ${{ secrets.DOCKERHUB_PASSWORD }}
+
+    - name: Push snapshot image to Docker Hub
+      if: ${{ github.ref == 'refs/heads/develop' }}
+      run: ./godelw docker push --tags=snapshot
+
+    - name: Push release image to Docker Hub
+      if: ${{ startsWith(github.ref, 'refs/tags/') }}
+      run: ./godelw docker push --tags=latest,version
+
+    - name: Find release
+      if: ${{ startsWith(github.ref, 'refs/tags/') }}
+      id: find-release
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      run: |
+        curl -sS --fail -H "Authorization: Bearer ${GITHUB_TOKEN}" "${GITHUB_API_URL}/repos/${GITHUB_REPOSITORY}/releases/tags/${GITHUB_REF#refs/tags/}" \
+          | jq .upload_url \
+          | sed 's/{.*}$//' \
+          | xargs -0 printf "::set-output name=upload_url::%s\n"
+
+    - name: Upload release
+      if: ${{ startsWith(github.ref, 'refs/tags/') }}
+      env:
+        UPLOAD_URL: ${{ steps.find-release.outputs.upload_url }}
+      run: |
+        dist_file=$(./godelw artifacts dist)
+        curl -sS --fail -X POST \
+          -H "Authorization: Bearer ${GITHUB_TOKEN}" \
+          -H "Content-Type: application/gzip" \
+          "${UPLOAD_URL}?name=$(basename ${dist_file})" \
+          --data-binary @${dist_file}
```

Comment:
I feel kind of bad just doing this with `curl` but the semi-official action to upload release assets is archived/deprecated and most community ones I found seem to also want to create the release. I guess a lot of people want to push a tag and have that automate the release creation instead of creating the release as a way to create the tag.

Another option is to write a new publisher plugin for Godel that can do this and the previous step. That's probably the best long-term option, but I wasn't ready to write that.

I know @jamestoyer likes `goreleaser`, but it looked like it also handles a bunch of the build process that we currently use Godel for and also works on the "push a tag to create a release" model.

Reply:
`goreleaser` looks nice. Autorelease Bot also follows this workflow of Create Release/Tag → Run CI to add artifacts/dists to the release, right?

Reply:
Yeah, Autorelease does create a tag before the release and then the tag creation triggers CI, but most (all?) of these projects don't actually attach artifacts to the release, but publish to other systems.

I was most flagging the workflow differences for people creating releases. Our current process is:

1. Publish a GitHub Release with release notes and a tag that does not exist yet
2. GitHub creates the tag
3. The tag triggers CI
4. CI publishes/upload artifacts

It seems like many of the open-source tools I found assume this process:

1. Locally create a tag and push it to GitHub
2. The tag triggers CI/automation
3. The automation creates a GitHub Release for the the existing tag
4. CI publishes/uploads artifacts

Reply:
A lot of OSS tooling will also be fine with an existing release. They will just update that. Generally, I much prefer using the `release` event to trigger the actual building of a release artifact